oak_functions/loader/src/lib.rs

```diff
@@ -25,3 +25,314 @@ pub mod metrics;
+    ///
+    /// Defaults to 4 if unset.
+    ///
+    /// Note that the CPU core detection logic does not seem to work reliably on Google Cloud Run,
+    /// so it is advisable to set this value to the number of cores available on the Cloud Run
+    /// instance.
+    ///
+    /// See <https://docs.rs/tokio/1.5.0/tokio/runtime/struct.Builder.html#method.worker_threads>.
+    worker_threads: Option<usize>,
+    /// Security policy guaranteed by the server.
+    policy: Option<Policy>,
+    /// Configuration for TensorFlow model
+    #[cfg(feature = "oak-tf")]
+    #[serde(default)]
+    tf_model: Option<TensorFlowModelConfig>,
+    /// Differentially private metrics configuration.
+    #[cfg(feature = "oak-metrics")]
+    #[serde(default)]
+    metrics: Option<PrivateMetricsConfig>,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+enum Data {
+    /// Download data file via HTTP GET.
+    /// Supported URL schemes: `http`, `https`.
+    Url(String),
+    /// Read data file from the local file system.
+    /// File path is relative to the current `$PWD` (*not* relative to the config file).
+    File(String),
+}
+
+/// Command line options for the Oak loader.
+///
+/// In general, when adding new configuration parameters, they should go in the `Config` struct
+/// instead of here, and provided as part of the config TOML file by the developer, who would
+/// normally bundle it with the Docker image of the Oak Functions Loader.
+#[derive(Parser, Clone, Debug)]
+#[clap(about = "Oak Functions Loader")]
+pub struct Opt {
+    #[clap(
+        long,
+        default_value = "8080",
+        help = "Port number that the server listens on."
+    )]
+    http_listen_port: u16,
+    #[clap(
+        long,
+        help = "Path to a Wasm file to be loaded and executed per invocation. The Wasm module must export a function named `main`."
+    )]
+    wasm_path: String,
+    #[clap(
+        long,
+        help = "Path to a file containing configuration parameters in TOML format."
+    )]
+    config_path: String,
+}
+
+async fn background_refresh_lookup_data(
+    lookup_data: &LookupData,
+    period: Duration,
+    logger: &Logger,
+) {
+    // Create an interval that starts after `period`, since the data was already refreshed
+    // initially.
+    let mut interval = tokio::time::interval_at(tokio::time::Instant::now() + period, period);
+    loop {
+        interval.tick().await;
+        // If there is an error, we skip the current refresh and wait for the next tick.
+        if let Err(err) = lookup_data.refresh().await {
+            logger.log_public(
+                Level::Error,
+                &format!("error refreshing lookup data: {}", err),
+            );
+        }
+    }
+}
+
+/// This crate is just a library so this `main` function does not get executed directly by anything,
+/// it needs to be wrapped in the "actual" main from a bin crate.
+pub fn main() -> anyhow::Result<()> {
```

Comment:
nit: to avoid confusion (e.g. by someone ignoring the comments above) perhaps it would be useful to rename to something like `lib_main` or `runtime_main` or something similar.

Reply:
Good idea, done.